### ISSUE TITLE
Crash when displaying U:\dev\zero

### ISSUE THREAD

**Issue by vinriviere** [freemint/freemint#183]
I'm not sure which component should be blamed here.
- With TeraDesk, go to U:\dev
- Double-click the `zero` file, and click Show
- And see what happens. At your convenience:
  - TeraDesk termination
  - FreeMiNT kernel panic
  - ARAyM termination

**Comment by mikrosk** [comment]
Maybe related to #174.

**Comment by DavidGZ** [comment]
With the CT60 Falcon Teradesk is killed, with the FireBee nothing from above happens, empty viewer window is shown, it seems the normal behaviour. Main difference here between both setups is that the kernel in the FireBee runs without memory protection.

**Comment by skaftetryne** [comment]
Try copy /dev/zero to a disk using Teradesk. /dev/zero has no EOF so Teradesk will continue copying until the destination disk is full. Probably the same problem with displaying it. It probably "works" without MP just because the buffer overrun is not detected right there and then.

**Comment by th-otto** [comment]
> /dev/zero has no EOF so Teradesk will continue copying until the destination disk is full.

Yes, that's normal behaviour of /dev/zero. Maybe some check should be added to Teradesk, not to attempt to copy devices files. But its more a missing feature than a bug.

That the same operation is able to kill mint, or even aranym, is another thing... killing mint might be a bug in teradesk (assuming that MP is off). Killing aranym is theoretically also possibly caused by this (assuming that some mint structures have been corrupted, and are passed to aranym by eg the XHDI interface, or the hostfs driver).

**Comment by mikrosk** [comment]
Shouldn't /dev/zero have something like 1 byte length or so?

**Comment by th-otto** [comment]
No, device files don't have a size. If it refers to some block device, you need some device specific ioctl()/fcntl to obtain the size (if any).

**Comment by th-otto** [comment]
I haven't found yet why teradesk crashes, but this is suely not a bug of mint. YOu can get similar resault by draging /dev/zero to qed (or some other viewer), which then reads from it until it runs out of memory.

Conclusion: just don't do that ;)

PS.: i'm not in the mood of trying it, but i would bet you can also panic a linux kernel by running an editor on /dev/zero

**Comment by ragnar76** [comment]
Hi there,

Am 27.06.20 um 11:51 schrieb Thorsten Otto:
> PS.: i'm not in the mood of trying it, but i would bet you can also
> panic a linux kernel by running an editor on /dev/zero
i was in the mood and tested it under my Linux Mint this is what i get:

- joe ; nothing happens. eats a lot of cpu time but system remains
responsible
- gnu nano ; print "this is not a regular file"
- vi ; prints "/dev/zero is not a file"
- mcedit (midnight commander) ; prints "/dev/zero is not a regular file"

bye
Bernd

**Comment by th-otto** [comment]
Thank you for testing it ;)

I would think joe remains responsible until it need to swap. When the swap space is also filled, it will either abort, or leave the system in an unusable state.

But the point here is: it would be the task of the viewer/editor to check for regular files, like the other editors do.

Of course you can't expect that from old programs that are not maintained anymore, but we should be able to change the internal viewer, qed, and teradesk at least.

**Comment by mikrosk** [comment]
In that case, happy to close this one. ;)

**Comment by vinriviere** [comment]
> In that case, happy to close this one. ;)

If you think that it is normal that **a user program** can **panic a kernel** and **exit an emulator**, then  go ahead, close this. But don't complain if people say that FreeMiNT is crap.

**Comment by mikrosk** [comment]
Well, it an user linux program (`joe`, as stated above) can make Linux kernel go out of swap and kernel panic, I am quite fine with that.

Having said that, adding a check to `qed` (at least, as it is our application) is definitely a wise thing to do.

**Comment by mfro0** [comment]
> If you think that it is normal that a user program can panic a kernel and exit an emulator, then go ahead, close this. But don't complain if people say that FreeMiNT is crap.

+1

There is no kernel panic with `joe /dev/zero` in Linux (at least not for me, I just tried out of curiosity).

The kernel just aborts the offending user process that tries to allocate more memory than available:

>     *** JOE was aborted due to swap file I/O error

To my knowledge, Linux reserves a certain amount of memory for the kernel to still be able to satisfy its memory demands even if a user program goes wild. I would expect MiNT doing something similar in this case.

**Comment by mikrosk** [comment]
Fair enough. ;)

**Comment by skaftetryne** [comment]
Tested on Milan060, Falcon/CT60 and Falcon/AB. Teradesk is killed due to a memory violation, most likely because of a buffer overrun (/dev/zero has no size, Teradesk allocates a small/zero buffer but reads indefinitely from /dev/zero when attempting to display it). But the rest of the system remains intact, and I can restart the desktop from the XaAES menu.

Without MP such a bug could definitely crash the system, which makes this a very good example of why MP exists in the first place.

> **exit an emulator**,

This only shows that the emulator does not emulate the real hardware close enough. Perfectly OK for Aranym I think, as it does not pretend to accurately emulate an existing machine.

Does this also happen with Aranym MMU and MP enabled?

**Comment by th-otto** [comment]
> Perfectly OK for Aranym I think, as it does not pretend to accurately emulate an existing machine.

Well i agree with Vincent here. Of course, without memory protection, that can easily kill the emulated system (there are lots of places where GEMDOS/AES  are missing checks on passed arguments, which would have similar effects). But (in theory) that should not be able to kill the emulator itself (unless the host does not have MP either, but that is not the case).

However fixing that in aranym is hardly possible, without accepting a large degree in emulation speed, just for the (very uncommon) case of such situations.

Adding a check in mint though, to return out-of-memory to applications *before* all the memory has been exhausted, might be an option. If that helps in this case is questionable though, there aren't that many places in the kernel where it would need some dynamically allocated memory, once the system has booted up.

So the best we can do is to fix Teradesk/qed, as these are the ones that are used for the snapshot archives. That might mean of course, if someone has configured his/her own viewer, the system may still crash, because that application cannot be fixed.

**Comment by skaftetryne** [comment]
> Adding a check in mint though, to return out-of-memory to applications _before_ all the memory has been exhausted, might be an option. If that helps in this case is questionable though,

I don't think that's the problem here. Teradesk is killed immediately - I suspect that it allocates a tiny (or zero-length) buffer for the file to display but quickly writes past this buffer as /dev/zero will never return EOF. Without MP it will simply overwrite everything from the start of this buffer to the end of the address range with zeros. What if Aranym crashes because Teradesk writes zeros to all the hardware registers?

IMO this is an Aranym-problem, not a MiNT-problem. On the contrary, with MP enabled MiNT handles this situation very well. Without MP... Well, what do you expect. Then there are no mechanisms to catch this Teradesk bug.
